### DISCUSSION TITLE
Can't access new.raw_app_meta_data inside a trigger

### DISCUSSION THREAD

**Discussion by evrazi** [supabase/supabase#19106] [Questions] [answered]
Want to save data into public.users table whenever a new user signs up. But I always get NULL if I use new.raw_app_meta_data ->> "role", which is a deal breaking since i want to use custom-claims.

CREATE OR REPLACE FUNCTION public.handle_new_user() 
RETURNS TRIGGER AS $$
BEGIN

  INSERT INTO public.users (id, role)
  VALUES (new.id, new.raw_app_meta_data->>'role');

  RETURN new;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;

CREATE TRIGGER on_auth_user_created
AFTER INSERT ON auth.users
FOR EACH ROW
EXECUTE PROCEDURE public.handle_new_user(); 

also tried which works without any problems.

BEGIN

  INSERT INTO public.users (id, role)
  VALUES (new.id, new.raw_user_meta_data->>'role'); //change app to user

  RETURN new;
END;

I am adding both users with 

`
supabase.auth.admin.createUser({
            email: 'email@email.com',
            password: 'password',
            email_confirm: true,
            app_metadata: { role: '1' }
          }) 
`

**Comment by GaryAustin1** [comment, marked as answer]
You may need to use an update trigger to get the raw_app_meta_data.   I believe there is a followup update on auth.users for that data.

**Reply by evrazi** [reply]
Something like this

First trigger is for inserting a new user
`
CREATE OR REPLACE FUNCTION public.handle_new_user() 
RETURNS TRIGGER AS $$
BEGIN
  -- Check if role is defined in raw_user_meta_data

  -- Insert into public.users
  INSERT INTO public.users (id)
  VALUES (new.id);

  RETURN new;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;


CREATE TRIGGER on_auth_user_created
AFTER INSERT ON auth.users
FOR EACH ROW
EXECUTE PROCEDURE public.handle_new_user();
`


Second one is for updating the user that was just created
`
CREATE OR REPLACE FUNCTION public.handle_update_app_meta_data() 
RETURNS TRIGGER AS $$
BEGIN

  UPDATE public.users
  SET role = new.raw_app_meta_data ->>"1"
  WHERE id = new.id;

  RETURN new;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;

CREATE TRIGGER on_auth_user_updated
AFTER Update ON auth.users
FOR EACH ROW
EXECUTE PROCEDURE public.handle_update_app_meta_data();
`

**Reply by GaryAustin1** [reply]
`SET role = new.raw_app_meta_data ->>'role'` ?

Also you can do `raise log 'new = %',new;` to see what is being passed in.  Look in the Postgres log in the dashboard.

**Reply by evrazi** [reply]
My mistake, wrote it as a reply first and then went to test it later.

 I am wondering now, are there are any recognizable or important differences between saving this data into `raw_user_meta_data` or `raw_app_meta_data`. Since `custom-claims` can be modified, but this can't work without two triggers. Is it okey if I ask it here or should I open a new discussion?

**Reply by GaryAustin1** [reply]
You cannot use raw_user_meta_data for anything secure.  The user can change it at anytime if they have a signed in jwt by using an updateUser call.
